### ISSUE TITLE
React Router 6.x prep/refactoring

### ISSUE THREAD

**Issue by ToppleTheNun** [WoWAnalyzer/WoWAnalyzer#5404]
I've been messing around with react-router 6.x and it's fairly neat. There are a few APIs that are currently in use that are no longer present in 6.x that would need to be refactored.

The intent of this issue isn't so much to fully advocate for moving to react-router 6.x _right now_, but more to facilitate doing some refactoring in order to smooth out the process of updating/upgrading should that be something we wish to do.

Below are some of the refactors that need to occur (some already have PRs put in for them):
- [x] `withRouter` is removed in favor of hooks
  - [x] App (#5399)
  - [x] NameSearch (#5400)
  - [x] ReportLoader (#5401)
  - [x] FightSelection (#5403)
  - [x] PlayerLoader (#5405)
  - [x] Results (#5408, #5409, #5412, #5425)
- [x] `render` prop on `Route` is removed in favor of `element` (can use `component` as the equivalent in 5.x)
  - [x] App (#5426)
    - [x] CharacterPage
    - [x] GuildPage
  - [x] Home (#5413)
    - [x] NewsPage
    - [x] ContributorPage
    - [x] Search

**Comment by ToppleTheNun** [comment]
Out of the original list, the `Results` refactor to use hooks is the most daunting due to the use of the legacy React context API. Fortunately, `updateResults` is only used in one place currently, but `contextTypes` is used in many more places. I might break that down into more PRs now that the `CombatLogParser` is available in a non-legacy context.

**Comment by ToppleTheNun** [comment]
# react-router v6

## Current State

I have a version of WoWAnalyzer working locally that has been completely converted to using react-router v6. It's not taking full advantage of the benefits of react-router v6 - [loaders](https://reactrouter.com/en/main/start/overview#data-loading) being the primary benefit that I would like to use, as I could see it reducing the complexity of the `Results` component structure. `Outlet`s would also be useful for the `Home` page specifically.

While a prototype is done, I can't quite recommend moving to react-router v6 yet due to some unfortunate API changes. The most outstanding issue is that react-router v6 does not currently support optional path parameters. This means that there are significant structural/function changes that need to occur with a fair amount of duplication.

The good news is that the team behind react-router have acknowledged that this is a gap and [have optional path parameters on their roadmap](https://togithub.com/remix-run/react-router/issues/9546).

## Cool Things

### Routing

If we take advantage of some of the cool things that react-router v6 brings like `Outlet`, then we can co-locate almost all of our routing. The below example makes it so that when you visit `/`, it renders the `Home` component and then inserts the `News` component into the rendered `Outlet`.

```tsx
// src/interface/Home.tsx
const Home = () => {
  // ...
  return (
    // ...
    <ErrorBoundary>
      <Outlet />
    </ErrorBoundary>
  );
}

// src/interface/App.tsx
const App = () => {
  // ...
  return (
    // ...
    <Routes>
      <Route path="/" element={<HomePage />}>
        <Route path="" element={<News />} />
        <Route path="news">
          <Route path=":articleId" element={<NewsPage />} />
          <Route path="" element={<News />} />
        </Route>
        <Route path="specs" element={<SpecList />} />
        <Route path="premium" element={<Premium />} />
        <Route path="about" element={<AboutPage />} />
        <Route path="help-wanted" element={<HelpWanted />} />
        <Route path="contributor/:id" element={<ContributorPage />} />
        <Route path="search">
          <Route path=":searchTerm" element={<Search />} />
          <Route path="" element={<Search />} />
        </Route>
        <Route path="*" element={<NotFound />} />
      </Route>
    </Routes>
  );
}
```

### Loaders

Loaders could certainly be useful. I haven't gotten to fully experiment with them as much as I'd like to, so I'll keep messing with them and post again here.

It would end up looking something like this, which is a relatively significant reduction in comparison to the current loader system:
```tsx
// src/interface/App.tsx
const App = () => {
  // ...
  return (
    // ...
    <Routes>
      <Route
        path="/report/:reportCode"
        loader={async ({ params }) => {
          const report = await fetchFights(params.reportCode ?? '');
          return defer({report});
        }}
        errorElement={<ErrorBoundary />}
      >
        <Route path=":fightId" element={<ReportPage />}>
        <Route path="" element={<ReportPage />} />
      </Route>
    </Routes>
  );
};
```
```tsx
// src/interface/report/ReportLoader.tsx
const ReportLoader = () => {
  const { report } = useLoaderData(); // this uses the closest loader to where this is rendered
  return (
    <Suspense
      fallback={
        <ActivityIndicator
          text={t({
            id: "interface.report.reportLoader",
            message: `Pulling report info...`,
          })}
        />
      }
    >
      <Await resolve={report}>
        {(resolvedReport) => (
          <>
            <Helmet>
              <title>{resolvedReport.title}</title>
            </Helmet>

            <ReportProvider
              report={resolvedReport}
              refreshReport={handleRefresh}
            >
              {children}
            </ReportProvider>
          </>
        )}
      </Await>
    </Suspense>
  );
};

```

## Issues

### `getMatch`

`getMatch` needs to become 5 functions, each with slightly different paths to match: `/report/:reportCode/*` becomes `/report/:reportCode/:fightId/*` and so on, with the `/*` being necessary to make it check against the entire path.
```ts
import { matchPath } from 'react-router-dom';

export const getMatchWithReportCode = (pathname: string) =>
  matchPath('/report/:reportCode/*', pathname);
```

### Routing

Because optional path parameters aren't a thing, what was previously a single `Route` becomes very duplicated.

**Before**
```tsx
const App = () => {
  // ...
  return (
    // ...
    <Switch>
      <Route
        path="/report/:reportCode?/:fightId?/:player?/:resultTab?"
        component={ReportPage}
      />
    </Switch>
  );
}
```

**After**
```tsx
const App = () => {
  // ...
  return (
    // ...
    <Routes>
      <Route path="/report/:reportCode">
        <Route path=":fightId" element={<ReportPage />}>
          <Route path=":player" element={<ReportPage />}>
            <Route path=":build" element={<ReportPage />}>
              <Route path=":resultTab" element={<ReportPage />}>
                <Route path="" element={<ReportPage />} />
              </Route>
              <Route path="" element={<ReportPage />} />
            </Route>
            <Route path="" element={<ReportPage />} />
          </Route>
          <Route path="" element={<ReportPage />} />
        </Route>
        <Route path="" element={<ReportPage />} />
      </Route>
    </Routes>
  );
}
```

## Demo

![react-router-v6](https://user-images.githubusercontent.com/1672786/203218852-dd7f4111-8417-4f1b-9d5e-6d5c50466ac7.gif)

**Comment by ToppleTheNun** [comment]
The use of `createBrowserRouter` and `RouterProvider router={browserRouter} />` (necessary to use loaders) instead of providing our own `BrowserRouter` has uncovered an issue where the `SupportChecker` modal was only rendering due to the location of a `getOpenModals` call.

This was originally in `App` and has been moved as close as possible to the source of the issue: I moved it into `Modal.tsx`, but the call itself is doing nothing but forcing a rerender. This solves the issue for both `SupportChecker` and for `ContributorButton`, which are the only places using a Modal currently.
```tsx
  useWaSelector((state) => getOpenModals(state));
```
